File: Ch05/solution/sizes.go
package main

import (
	"fmt"
	"os"
)

type response struct {
	fileName string
	size     int64
}

func worker(fileName string, ch chan<- response) {
	resp := response{fileName: fileName, size: -1}
	st, err := os.Stat(fileName)
	if err == nil {
		resp.size = st.Size()
	}
	ch <- resp
}

func main() {

	files := []string{
		"files/a.txt",
		"files/b.txt",
		"files/c.txt",
		"files/d.txt",
		"files/e.txt",
	}
	ch := make(chan response)

	for _, f := range files {
		go worker(f, ch)
	}

	for range files {
		r := <-ch
		fmt.Printf("%s -> %d\n", r.fileName, r.size)
	}
}

File: Ch05/solution/sizes.py
from pathlib import Path
from queue import Queue
from threading import Thread


def worker(file_name, queue):
    size = Path(file_name).stat().st_size
    queue.put((file_name, size))


files = [
    'Ch05/challenge/files/a.txt',
    'Ch05/challenge/files/b.txt',
    'Ch05/challenge/files/c.txt',
    'Ch05/challenge/files/d.txt',
    'Ch05/challenge/files/e.txt',
]

queue = Queue()
for file_name in files:
    Thread(target=worker, args=(file_name, queue), daemon=True).start()

for _ in files:
    file_name, size = queue.get()
    print(f'{file_name} -> {size}')
